pandas/core/arrays/interval.py

```diff
@@ -0,0 +1,814 @@
+    Copy the meta-data.
+dtype : dtype or None, default None
+    If None, dtype will be inferred
+
+    .. versoinadded:: 0.23.0
+
+Attributes
+----------
+left
+right
+closed
+mid
+length
+values
+is_non_overlapping_monotonic
+
+Methods
+-------
+from_arrays
+from_tuples
+from_breaks
+%(extra_methods)s\
+
+%(examples)s\
+
+Notes
+------
+See the `user guide
+<http://pandas.pydata.org/pandas-docs/stable/advanced.html#intervalindex>`_
+for more.
+
+See Also
+--------
+Index : The base pandas Index type
+Interval : A bounded slice-like interval; the elements of an IntervalIndex
+interval_range : Function to create a fixed frequency IntervalIndex
+cut, qcut : Convert arrays of continuous data into Categoricals/Series of
+            Intervals
+"""
+
+
+@Appender(_interval_shared_docs['class'] % dict(
+    klass="IntervalArray",
+    summary="Pandas array for interval data that are closed on the same side",
+    versionadded="0.23.0",
+    name='', extra_methods='', examples='',
+))
+@add_metaclass(_WritableDoc)
+class IntervalArray(IntervalMixin, ExtensionArray):
+    dtype = IntervalDtype()
+    ndim = 1
+    can_hold_na = True
+    _na_value = _fill_value = np.nan
+
+    def __new__(cls, data, closed=None, dtype=None, copy=False,
+                fastpath=False, verify_integrity=True):
+
+        if fastpath:
+            return cls._simple_new(data.left, data.right, closed,
+                                   copy=copy, dtype=dtype,
+                                   verify_integrity=False)
+
+        if isinstance(data, ABCSeries) and is_interval_dtype(data):
+            data = data.values
+        if isinstance(data, (cls, ABCIntervalIndex)):
+            left = data.left
+            right = data.right
+            closed = data.closed
+        else:
+
+            # don't allow scalars
+            if is_scalar(data):
+                msg = ("{}(...) must be called with a collection of some kind,"
+                       " {} was passed")
+                raise TypeError(msg.format(cls.__name__, data))
+
+            data = maybe_convert_platform_interval(data)
+            left, right, infer_closed = intervals_to_interval_bounds(data)
+
+            if (com._all_not_none(closed, infer_closed) and
+                    closed != infer_closed):
+                # GH 18421
+                msg = ("conflicting values for closed: constructor got "
+                       "'{closed}', inferred from data '{infer_closed}'"
+                       .format(closed=closed, infer_closed=infer_closed))
+                raise ValueError(msg)
+
+            closed = closed or infer_closed
+
+        return cls._simple_new(left, right, closed, copy=copy, dtype=dtype,
+                               verify_integrity=verify_integrity)
+
+    @classmethod
+    def _simple_new(cls, left, right, closed=None,
+                    copy=False, dtype=None, verify_integrity=True):
+        result = IntervalMixin.__new__(cls)
+
+        closed = closed or 'right'
```

Comment:
did we default this before here? (e.g. this should be filled in I think by now). maybe assert that it is not None here (and push the default higher up). IOW don't want to have in too many places.

Reply:
There are a couple corner cases: empty data and purely NA data.  In both cases, the inferred `closed` will be `None`, so if `closed` is also not passed to the constructor it will be `None` here.